api/src/main/java/org/openmrs/module/sync/api/impl/SyncServiceImpl.java

```diff
@@ -296,50 +296,59 @@ public List<SyncRecord> getSyncRecords(SyncRecordState[] states, Integer maxSync
 	 * @see org.openmrs.module.sync.api.SyncService#getSyncRecords(org.openmrs.module.sync.SyncRecordState[],
 	 *      org.openmrs.module.sync.server.RemoteServer, java.lang.Integer, java.lang.Integer)
 	 */
-	@Transactional(readOnly = true)
-	public List<SyncRecord> getSyncRecords(SyncRecordState[] states, RemoteServer server, Integer maxSyncRecords, Integer firstRecordId)
-	                                                                                                             throws APIException {
-		List<SyncRecord> temp = null;
-		List<SyncRecord> ret = null;
-		
-		if (server != null) {
+	public List<SyncRecord> getSyncRecordsAndUpdateAsNotSupposedToSync(SyncRecordState[] states,
+																	   RemoteServer server,
+																	   Integer maxSyncRecords,
+																	   Integer firstRecordId) throws APIException {

```

Comment:
The crux of the below change is that before, we were getting X sync records, and then once we had this limited set of records, we were iterating over it, marking those as Not Supposed To Sync that qualified, and then returning the List of those that were supposed to sync.  This meant that the list returned was often much smaller (frequently empty) than the maxSyncRecords even when many records were pending to sync.

So the change I made was to retrieve the maxSyncRecords number of sync records iteratively, removing any as Not Supposed To Sync along the way, and continuing to add more iterations of records until the maxSyncRecords number is hit, containing all records that are supposed to sync.